### PR TITLE
community: Allow other than default parsers in SharePointLoader and OneDriveLoader

### PR DESCRIPTION
**Pull request by MacanPN**
### Currently `O365BaseLoader` (and consequently both derived loaders) are limited to `pdf`, `doc`, `docx` files.
- **Solution: here we introduce _handlers_ attribute that allows for custom handlers to be passed in. This is done in _dict_ form:**

**Example:**
```python
from langchain_community.document_loaders.parsers.documentloader_adapter import DocumentLoaderAsParser
# PR for DocumentLoaderAsParser here: https://github.com/langchain-ai/langchain/pull/27749
from langchain_community.document_loaders.excel import UnstructuredExcelLoader

xlsx_parser = DocumentLoaderAsParser(UnstructuredExcelLoader, mode="paged")

# create dictionary mapping file types to handlers (parsers)
handlers = {
    "doc": MsWordParser()
    "pdf": PDFMinerParser()
    "txt": TextParser()
    "xlsx": xlsx_parser
}
loader = SharePointLoader(document_library_id="...",
                            handlers=handlers # pass handlers to SharePointLoader
                            )
documents = loader.load()

# works the same in OneDriveLoader
loader = OneDriveLoader(document_library_id="...",
                            handlers=handlers
                            )
```
This dictionary is then passed to `MimeTypeBasedParser` same as in the [current implementation](https://github.com/langchain-ai/langchain/blob/5a2cfb49e045988d290a1c7e3a0c589d6b371694/libs/community/langchain_community/document_loaders/parsers/registry.py#L13). 

### Currently `SharePointLoader` and `OneDriveLoader` are separate loaders that both inherit from `O365BaseLoader`
However both of these implement the same functionality. The only differences are:
  - `SharePointLoader` requires argument `document_library_id` whereas `OneDriveLoader` requires `drive_id`. These are just different names for the same thing.
  - `SharePointLoader` implements significantly more features.
- **Solution: `OneDriveLoader` is replaced with an empty shell just renaming `drive_id` to `document_library_id` and inheriting from `SharePointLoader`**

**Dependencies:** None
**Twitter handle:** @martintriska1

If no one reviews your PR within a few days, please @-mention one of baskaryan, efriis, eyurtsev, ccurme, vbarda, hwchase17.

### CODE REVIEW

File: libs/community/langchain_community/document_loaders/base_o365.py
Comment (vbarda) [inline libs/community/langchain_community/document_loaders/base_o365.py:55]: could we just use a simple dictionary lookup instead?

Comment (vbarda) [reply libs/community/langchain_community/document_loaders/base_o365.py:55]: (and same for below)

File: libs/community/langchain_community/document_loaders/base_o365.py
Comment (vbarda) [inline libs/community/langchain_community/document_loaders/base_o365.py:86]: for the keys (file types) let's either keep the original `FileType` enum, or use a `Literal["doc", "pdf", "txt"]` for the annotation

File: libs/community/langchain_community/document_loaders/onedrive.py
Comment (vbarda) [inline libs/community/langchain_community/document_loaders/onedrive.py:76]: is this removed because the one in `SharePointLoader` is more comprehensive?

File: libs/community/langchain_community/document_loaders/onedrive.py
Comment (vbarda) [inline libs/community/langchain_community/document_loaders/onedrive.py:9]: why remove `Microsoft`?

**Review by vbarda — COMMENTED** [review]
Thanks for your contribution! 

Looks good overall, but my biggest question is: can't we just allow a single optional `MimeTypeBasedParser` as an input to this document loader? I think it could simplify a lot of this logic. The user could then do something like:

```python
blob_parser = MimeTypeBasedParser({
    "application/msword": MsWordParser(),
    "application/pdf": PDFMinerParser(),
    ""audio/mpeg": OpenAIWhisperParser()
})
loader = OneDriveLoader(..., blob_parser=blob_parser)
```

File: libs/community/langchain_community/document_loaders/base_o365.py
Comment (vbarda) [reply libs/community/langchain_community/document_loaders/base_o365.py:86]: (pending the higher level comment about simply passing a parser)

File: libs/community/langchain_community/document_loaders/onedrive.py
Comment (vbarda) [reply libs/community/langchain_community/document_loaders/onedrive.py:76]: (saw you mention this in the PR description)

File: libs/community/langchain_community/document_loaders/base_o365.py
Comment (MacanPN) [reply libs/community/langchain_community/document_loaders/base_o365.py:55]: Who's going to maintain that lookup dict? Under the hood that is exactly what `mimetypes` does, except that library _specializes_ in this so their lookup will be always more up to date and complete. Why not offload the conversion to a library?

File: libs/community/langchain_community/document_loaders/onedrive.py
Comment (MacanPN) [reply libs/community/langchain_community/document_loaders/onedrive.py:9]: I rewrote the whole comment and ended up leaving out "Microsoft". I can put that back.

**Comment by MacanPN** [comment]
> Thanks for your contribution!
> 
> Looks good overall, but my biggest question is: can't we just allow a single optional `MimeTypeBasedParser` as an input to this document loader? I think it could simplify a lot of this logic. The user could then do something like:
> 
> ```python
> blob_parser = MimeTypeBasedParser({
>     "application/msword": MsWordParser(),
>     "application/pdf": PDFMinerParser(),
>     ""audio/mpeg": OpenAIWhisperParser()
> })
> loader = OneDriveLoader(..., blob_parser=blob_parser)
> ```

This is a good point. I'll change the implementation so we're passing in `Parser` rather than `handlers` dictionary.

**Comment by MacanPN** [comment]
In general the extra logic implemented in this PR is not about instantiation of `MimeTypeBasedParser` (that is 1 line thing) but rather about supporting arbitrary file types (rather than just 3 enumerated types).

File: libs/community/langchain_community/document_loaders/base_o365.py
Comment (vbarda) [reply libs/community/langchain_community/document_loaders/base_o365.py:55]: fair point!

File: libs/community/langchain_community/document_loaders/base_o365.py
Comment (vbarda) [inline libs/community/langchain_community/document_loaders/base_o365.py:150]: let's import at the top of the file

**Review by vbarda — APPROVED** [review]

**Comment by vbarda** [comment]
> > Thanks for your contribution!
> > Looks good overall, but my biggest question is: can't we just allow a single optional `MimeTypeBasedParser` as an input to this document loader? I think it could simplify a lot of this logic. The user could then do something like:
> > ```python
> > blob_parser = MimeTypeBasedParser({
> >     "application/msword": MsWordParser(),
> >     "application/pdf": PDFMinerParser(),
> >     ""audio/mpeg": OpenAIWhisperParser()
> > })
> > loader = OneDriveLoader(..., blob_parser=blob_parser)
> > ```
> 
> This is a good point. I've been thinking about this but consider:
> 
> * We still need to implement functions for `file extension <-> mime type` conversions.
> * We still need to set up `_file_types` and `_mime_types` properties (for interface compatibility)
> * We still need to provide default parser.
>   So we wouldn't really simplify the logic that much. What the implementation has on top is option to pass in file types (file extensions) instead of mime types, which I think is a real quality of life improvement. Some mime types can be not as straight forward as the ones in the example above, such as:
> * `.docx`: `application/vnd.openxmlformats-officedocument.wordprocessingml.document`
> * `xlsx`: `application/vnd.openxmlformats-officedocument.spreadsheetml.sheet`
> 
> Please also consider that one could not simply pass in general `BaseBlobParser` but it would have to be `MimeTypeBasedParser` (because we are looking at `handlers` to determine whether the file should be downloaded and parsed or not).

Makes sense, thanks for the follow up! Approved pending two small nits. IMO this is the key argument in favor of this change:

> What the implementation has on top is option to pass in file types (file extensions) instead of mime types, which I think is a real quality of life improvement. Some mime types can be not as straight forward as the ones in the example above, such as:
> .docx: application/vnd.openxmlformats-officedocument.wordprocessingml.document
> xlsx: application/vnd.openxmlformats-officedocument.spreadsheetml.sheet

File: libs/community/langchain_community/document_loaders/onedrive.py
Comment (MacanPN) [reply libs/community/langchain_community/document_loaders/onedrive.py]: Done

File: libs/community/langchain_community/document_loaders/base_o365.py
Comment (MacanPN) [reply libs/community/langchain_community/document_loaders/base_o365.py]: Done

**Comment by MacanPN** [comment]
@vbarda Minor nitpicks addressed. Are we good to merge?

**Comment by MacanPN** [comment]
@vbarda I can't merge it myself. Can you please merge the PR?
image

**Comment by rasc0der** [comment]
Thanks for your contribution!

**Pull request merged**
